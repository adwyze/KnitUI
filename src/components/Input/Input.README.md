<!-- STORY -->

# Input Component

Usage:

```javascript
<div
  style={{
    width: 240,
  }}>
  <Input onChange={action("input changed")} placeholder="basic usage" />
</div>
```
