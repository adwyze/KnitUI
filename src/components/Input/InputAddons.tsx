import React, { ReactNode } from "react"
import _ from "lodash"

import { IInputProps } from "./types"
import { getIconSize } from "./utils"

import {
  AddonBeforeSpan,
  AddonAfterSpan,
  AddonContainer,
  StyledLabel,
  NotificationContainer,
} from "./styledComponents"

const renderInputAddons = (
  children: React.ReactElement<any>,
  props: IInputProps
) => {
  const {
    addonAfter,
    addonBefore,
    label,
    notification,
    inputSize,
    state,
  } = props
  const customProps = {
    inputSize,
    state,
  }

  let labelDOM: null | ReactNode = null
  let notificationDOM: null | ReactNode = null
  let addonDOM: ReactNode = children

  if (label) {
    labelDOM = label
    if (_.isString(label)) {
      labelDOM = <StyledLabel customProps={customProps}>{label}</StyledLabel>
    }
  }

  if (notification) {
    notificationDOM = notification
    if (_.isString(notification)) {
      notificationDOM = (
        <NotificationContainer customProps={customProps}>
          {notification}
        </NotificationContainer>
      )
    }
  }

  if (addonBefore) {
    addonDOM = (
      <AddonBeforeSpan iconSize={getIconSize(inputSize)}>
        {addonBefore ? (
          <AddonContainer
            css={`
              left: 1rem;
            `}>
            {addonBefore}
          </AddonContainer>
        ) : null}
        {children}
      </AddonBeforeSpan>
    )
  }
  if (addonAfter) {
    addonDOM = (
      <AddonAfterSpan iconSize={getIconSize(inputSize)}>
        {children}
        {addonAfter ? (
          <AddonContainer
            css={`
              right: 1rem;
            `}>
            {addonAfter}
          </AddonContainer>
        ) : null}
      </AddonAfterSpan>
    )
  }
  return (
    <>
      {labelDOM}
      {addonDOM}
      {notificationDOM}
    </>
  )
}

export default renderInputAddons
