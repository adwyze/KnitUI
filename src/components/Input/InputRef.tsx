import React, { useState, useEffect, useRef } from "react"

import { IInputProps } from "./types"
import { getIconSize } from "./utils"

import RenderInput from "./RenderInput"
import renderInputAddons from "./InputAddons"
import { IconBackground, InputWrapper, AddonIcon } from "./styledComponents"

const InputRef = React.forwardRef<HTMLElement, IInputProps>((props, ref) => {
  const [currentValue, setCurrentValue] = useState<string | number | undefined>(
    props.value || ""
  )
  const [isPasswordField, setIsPasswordField] = useState(
    props.type === "password"
  )
  const [isOverflow, setIsOverflow] = useState(false)

  useEffect(() => {
    // To update the current input value
    setCurrentValue(props.value)
  }, [props.value])

  const currentRef = useRef<HTMLElement>(null) as React.MutableRefObject<
    HTMLElement
  >

  const customProps = {
    inputSize: props.inputSize,
    state: props.state,
  }

  const handleChange = e => {
    // Handle onChange if provided and update the input value
    if (props.onChange) {
      props.onChange(e)
    }

    setCurrentValue(e.target.value)
  }

  const handleFocus = e => {
    if (props.onFocus) {
      props.onFocus(e)
    }

    setIsOverflow(false)
  }

  const handleBlur = e => {
    if (props.onBlur) {
      props.onBlur(e)
    }

    if (currentRef.current) {
      const input = currentRef.current.children[0]

      if (input.scrollWidth > input.clientWidth && !isOverflow) {
        setIsOverflow(true)
      } else if (input.scrollWidth <= input.clientWidth && isOverflow) {
        setIsOverflow(false)
      }
    }
  }

  const clearInput = () => {
    setCurrentValue("")
  }

  const togglePasswordField = () => {
    setIsPasswordField(prevState => !prevState)
  }

  const getCurrentInputType = () => {
    // Get the input type. Useful while toggling password field.
    if (props.type === "number") {
      return "number"
    } else if (props.type === "textarea") {
      return "textarea"
    } else if (isPasswordField) {
      return "password"
    }
    return "default"
  }

  // Implemented to save the effort of adding `getIconSize` on every Icon component
  const Icon = iconProps => (
    <AddonIcon size={getIconSize(props.inputSize)} {...iconProps} />
  )

  const InputElem = (
    <InputWrapper
      inputSize={props.inputSize}
      type={props.type}
      state={props.state}
      addonAfter={props.addonAfter}
      ref={currentRef}
      customProps={customProps}
      isOverflow={isOverflow}>
      <RenderInput
        {...props}
        ref={ref}
        value={currentValue}
        onChange={handleChange}
        type={getCurrentInputType()}
        isOverflow={isOverflow}
        onBlur={handleBlur}
        onFocus={handleFocus}
      />
    </InputWrapper>
  )

  let addonAfter = props.addonAfter

  if (props.type === "search") {
    if (currentValue && currentValue.toString().length > 0) {
      addonAfter = (
        <IconBackground bgSize={getIconSize(props.inputSize)}>
          <Icon type="oClose" fill="#000000" onClick={clearInput} pointer />
        </IconBackground>
      )
    } else {
      addonAfter = <Icon type="oSearch" fill="#a6a6a6" />
    }
  } else if (props.type === "password") {
    if (isPasswordField) {
      addonAfter = (
        <Icon
          type="oVisibility"
          fill="#000000"
          onClick={togglePasswordField}
          pointer
        />
      )
    } else {
      addonAfter = (
        <Icon
          type="oVisibilityOff"
          fill="#000000"
          onClick={togglePasswordField}
          pointer
        />
      )
    }
  } else if (props.state === "success") {
    addonAfter = <Icon type="oCheckCircleOutline" fill="#036600" />
  } else if (props.state === "warning") {
    addonAfter = <Icon type="oWarning" fill="#FFC700" />
  } else if (props.state === "error") {
    addonAfter = <Icon type="oErrorOutline" fill="#990000" />
  }

  return renderInputAddons(InputElem as React.ReactElement<any>, {
    ...props,
    addonAfter,
  })
})

export default InputRef
