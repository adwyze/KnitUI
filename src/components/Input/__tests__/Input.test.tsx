import React from "react"
import Input from ".."
import { cleanup, render, fireEvent } from "../../../common/TestUtil"
import Icon from "../../Icon"

afterEach(cleanup)

describe("Input", () => {
  const sizes = ["small", "large", undefined]
  const types = ["search", "number", "password", "textarea", undefined]
  const states = ["success", "warning", "error", undefined]
  const resizes = ["horizontal", "vertical", "none", undefined]

  // All tests are executes for each of the three sizes
  sizes.forEach((size: "small" | "large" | undefined) => {
    describe(`with size ${size}`, () => {
      it("renders correctly", () => {
        const { asFragment } = render(
          <Input
            placeholder="placeholder"
            inputSize={size}
            onChange={() => {}}
          />
        )
        expect(asFragment()).toMatchSnapshot()
      })
    })
  })

  types.forEach(
    (type: "search" | "number" | "password" | "textarea" | undefined) => {
      describe(`with type ${type}`, () => {
        it("renders correctly", () => {
          const { asFragment } = render(
            <Input placeholder="placeholder" type={type} />
          )
          expect(asFragment()).toMatchSnapshot()
        })
      })
    }
  )

  states.forEach((state: "success" | "warning" | "error" | undefined) => {
    describe(`with state ${state}`, () => {
      it("renders correctly", () => {
        const { asFragment } = render(<Input state={state} />)
        expect(asFragment()).toMatchSnapshot()
      })
    })
  })

  resizes.forEach((resize: "horizontal" | "vertical" | "none" | undefined) => {
    describe(`with textarea resize: ${resize}`, () => {
      it("renders correctly", () => {
        const { asFragment } = render(
          <Input type="textarea" textareaResize={resize} />
        )
        expect(asFragment()).toMatchSnapshot()
      })
    })
  })

  it("with a text label", () => {
    const { asFragment } = render(
      <Input placeholder="placeholder" onChange={() => {}} label="label" />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("with a text notification", () => {
    const { asFragment } = render(
      <Input
        placeholder="placeholder"
        onChange={() => {}}
        notification="notification"
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("with a custom label", () => {
    const { asFragment } = render(
      <Input
        placeholder="placeholder"
        onChange={() => {}}
        label={
          <div
            css={`
              display: flex;
              align-items: center;
            `}>
            <span>Label</span>
            <Icon type="oInfo" />
          </div>
        }
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("with a custom notification", () => {
    const { asFragment } = render(
      <Input
        placeholder="placeholder"
        onChange={() => {}}
        notification={
          <div
            css={`
              display: flex;
              align-items: center;
            `}>
            <span>Notification</span>
            <Icon type="oInfo" />
          </div>
        }
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("with a text label and a text notification", () => {
    const { asFragment } = render(
      <Input
        placeholder="placeholder"
        onChange={() => {}}
        label="label"
        notification="notification"
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("with left addon renders correctly", () => {
    const { asFragment } = render(
      <Input
        placeholder="placeholder"
        onChange={() => {}}
        addonBefore={<Icon type="oInfo" />}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("with right addon renders correctly", () => {
    const { asFragment } = render(
      <Input
        placeholder="placeholder"
        onChange={() => {}}
        addonAfter={<Icon type="oInfo" />}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("with left and right addons renders correctly", () => {
    const { asFragment } = render(
      <Input
        placeholder="placeholder"
        onChange={() => {}}
        addonBefore={<Icon type="oInfo" />}
        addonAfter={<Icon type="oInfo" />}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("should apply provided classname", () => {
    const { container } = render(
      <Input
        placeholder="placeholder"
        onChange={() => {}}
        className="custom-class"
      />
    )

    if (container.firstChild) {
      expect(container.firstChild.firstChild).toHaveClass("custom-class")
    }
  })

  it("should apply provided styled", () => {
    const { container } = render(
      <Input
        placeholder="placeholder"
        onChange={() => {}}
        style={{ backgroundColor: "red" }}
      />
    )

    if (container.firstChild) {
      expect(container.firstChild.firstChild).toHaveStyle(
        "background-color: red"
      )
    }
  })

  it("should display disabled icon on hover for disabled input", () => {
    const { container } = render(<Input onChange={() => {}} disabled />)

    if (container.firstChild) {
      expect(container.firstChild.firstChild).toHaveStyle("cursor: not-allowed")
    }
  })

  it("should display search icon on type search", () => {
    const { container } = render(<Input type="search" />)

    const iconContainer = container.querySelector('span[type="oSearch"]')

    expect(iconContainer).toBeDefined()
  })

  it("should display clear icon on type search with value", () => {
    const { container } = render(<Input type="search" value="test" />)

    const iconContainer = container.querySelector('span[type="oClose"]')

    expect(iconContainer).toBeDefined()
  })

  it("should display show password icon on type password", () => {
    const { container } = render(<Input type="password" />)

    const input = container.querySelector('input[type="password"]')
    const iconContainer = container.querySelector('span[type="oVisibility"]')

    expect(iconContainer).toBeDefined()
    expect(input).toBeDefined()
  })

  it("should toggle password mode with clicked on icons", () => {
    const { container } = render(<Input type="password" />)

    const input = container.querySelector('input[type="password"]')
    const showPasswordContainer = container.querySelector(
      'span[type="oVisibility"]'
    )

    expect(input).toBeDefined()
    expect(showPasswordContainer).toBeDefined()

    if (input) {
      expect(input.getAttribute("type")).toEqual("password")
    }

    if (showPasswordContainer) {
      // Click on show password button
      fireEvent.click(showPasswordContainer)

      const hidePasswordContainer = container.querySelector(
        'span[type="oVisibilityOff"]'
      )
      // Expect hide password icon to exist
      expect(hidePasswordContainer).toBeDefined()

      if (input) {
        // Input type changed from password to text
        expect(input.getAttribute("type")).toEqual("text")
      }

      if (hidePasswordContainer) {
        // Click on hide password button
        fireEvent.click(hidePasswordContainer)
        // Expect show password button to exist
        expect(showPasswordContainer).toBeDefined()

        if (input) {
          // Input type changed from text to password
          expect(input.getAttribute("type")).toEqual("password")
        }
      }
    }
  })

  it("should display textarea", () => {
    const { container } = render(<Input type="textarea" />)

    if (container.firstChild && container.firstChild.firstChild) {
      expect(container.firstChild.firstChild.nodeName).toEqual("TEXTAREA")
    }
  })
})
