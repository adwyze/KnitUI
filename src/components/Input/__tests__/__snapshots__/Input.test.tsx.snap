// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input with a custom label 1`] = `
<DocumentFragment>
  .c0 {
  cursor: inherit;
  display: inline-block;
  height: 18px;
  width: 18px;
}

<div
    css="
              display: flex;
              align-items: center;
            "
  >
    <span>
      Label
    </span>
    <span
      class="c0"
      height="18px"
      type="oInfo"
      width="18px"
    >
      <svg
        fill="#000"
        height="18px"
        viewBox="0 0 24 24"
        width="18px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 7H13V9H11V7ZM11 11H13V17H11V11ZM12 2C6.48 2 2 6.48 2 12C2 17.52 6.48 22 12 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 12 2ZM12 20C7.59 20 4 16.41 4 12C4 7.59 7.59 4 12 4C16.41 4 20 7.59 20 12C20 16.41 16.41 20 12 20Z"
        />
      </svg>
    </span>
  </div>
  .c0 {
  height: 2.8rem;
  width: 100%;
  margin: 0.4rem 0;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 0.4rem 1.4rem;
  box-sizing: border-box;
  background-color: hsl(0,0%,95%);
  color: hsl(0,0%,5%);
  font-size: 1.4rem;
  line-height: 2rem;
}

.c0:hover {
  background-color: hsl(0,0%,90%);
  color: hsl(0,0%,5%);
}

.c0:focus,
.c0:active {
  outline: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid #5ab2f2;
  box-shadow: 0 0 2px hsl(216,100%,50%);
}

.c0::-webkit-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::-moz-placeholder {
  color: hsl(0,0%,60%);
}

.c0:-ms-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::placeholder {
  color: hsl(0,0%,60%);
}

.c0:disabled {
  cursor: not-allowed;
  background-color: hsl(0,0%,95%);
  border: 1px solid transparent;
  box-shadow: none;
}

<input
    class="c0"
    placeholder="placeholder"
  />
</DocumentFragment>
`;

exports[`Input with a custom notification 1`] = `
<DocumentFragment>
  .c0 {
  height: 2.8rem;
  width: 100%;
  margin: 0.4rem 0;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 0.4rem 1.4rem;
  box-sizing: border-box;
  background-color: hsl(0,0%,95%);
  color: hsl(0,0%,5%);
  font-size: 1.4rem;
  line-height: 2rem;
}

.c0:hover {
  background-color: hsl(0,0%,90%);
  color: hsl(0,0%,5%);
}

.c0:focus,
.c0:active {
  outline: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid #5ab2f2;
  box-shadow: 0 0 2px hsl(216,100%,50%);
}

.c0::-webkit-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::-moz-placeholder {
  color: hsl(0,0%,60%);
}

.c0:-ms-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::placeholder {
  color: hsl(0,0%,60%);
}

.c0:disabled {
  cursor: not-allowed;
  background-color: hsl(0,0%,95%);
  border: 1px solid transparent;
  box-shadow: none;
}

<input
    class="c0"
    placeholder="placeholder"
  />
  .c0 {
  cursor: inherit;
  display: inline-block;
  height: 18px;
  width: 18px;
}

<div
    css="
              display: flex;
              align-items: center;
            "
  >
    <span>
      Notification
    </span>
    <span
      class="c0"
      height="18px"
      type="oInfo"
      width="18px"
    >
      <svg
        fill="#000"
        height="18px"
        viewBox="0 0 24 24"
        width="18px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 7H13V9H11V7ZM11 11H13V17H11V11ZM12 2C6.48 2 2 6.48 2 12C2 17.52 6.48 22 12 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 12 2ZM12 20C7.59 20 4 16.41 4 12C4 7.59 7.59 4 12 4C16.41 4 20 7.59 20 12C20 16.41 16.41 20 12 20Z"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`Input with a text label 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1.2rem;
  line-height: 1.8rem;
  color: hsl(0,0%,5%);
}

<label
    class="c0"
  >
    label
  </label>
  .c0 {
  height: 2.8rem;
  width: 100%;
  margin: 0.4rem 0;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 0.4rem 1.4rem;
  box-sizing: border-box;
  background-color: hsl(0,0%,95%);
  color: hsl(0,0%,5%);
  font-size: 1.4rem;
  line-height: 2rem;
}

.c0:hover {
  background-color: hsl(0,0%,90%);
  color: hsl(0,0%,5%);
}

.c0:focus,
.c0:active {
  outline: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid #5ab2f2;
  box-shadow: 0 0 2px hsl(216,100%,50%);
}

.c0::-webkit-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::-moz-placeholder {
  color: hsl(0,0%,60%);
}

.c0:-ms-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::placeholder {
  color: hsl(0,0%,60%);
}

.c0:disabled {
  cursor: not-allowed;
  background-color: hsl(0,0%,95%);
  border: 1px solid transparent;
  box-shadow: none;
}

<input
    class="c0"
    placeholder="placeholder"
  />
</DocumentFragment>
`;

exports[`Input with a text label and a text notification 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1.2rem;
  line-height: 1.8rem;
  color: hsl(0,0%,5%);
}

<label
    class="c0"
  >
    label
  </label>
  .c0 {
  height: 2.8rem;
  width: 100%;
  margin: 0.4rem 0;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 0.4rem 1.4rem;
  box-sizing: border-box;
  background-color: hsl(0,0%,95%);
  color: hsl(0,0%,5%);
  font-size: 1.4rem;
  line-height: 2rem;
}

.c0:hover {
  background-color: hsl(0,0%,90%);
  color: hsl(0,0%,5%);
}

.c0:focus,
.c0:active {
  outline: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid #5ab2f2;
  box-shadow: 0 0 2px hsl(216,100%,50%);
}

.c0::-webkit-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::-moz-placeholder {
  color: hsl(0,0%,60%);
}

.c0:-ms-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::placeholder {
  color: hsl(0,0%,60%);
}

.c0:disabled {
  cursor: not-allowed;
  background-color: hsl(0,0%,95%);
  border: 1px solid transparent;
  box-shadow: none;
}

<input
    class="c0"
    placeholder="placeholder"
  />
  .c0 {
  font-size: 1.2rem;
  line-height: 1.8rem;
}

<div
    class="c0"
  >
    notification
  </div>
</DocumentFragment>
`;

exports[`Input with a text notification 1`] = `
<DocumentFragment>
  .c0 {
  height: 2.8rem;
  width: 100%;
  margin: 0.4rem 0;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 0.4rem 1.4rem;
  box-sizing: border-box;
  background-color: hsl(0,0%,95%);
  color: hsl(0,0%,5%);
  font-size: 1.4rem;
  line-height: 2rem;
}

.c0:hover {
  background-color: hsl(0,0%,90%);
  color: hsl(0,0%,5%);
}

.c0:focus,
.c0:active {
  outline: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid #5ab2f2;
  box-shadow: 0 0 2px hsl(216,100%,50%);
}

.c0::-webkit-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::-moz-placeholder {
  color: hsl(0,0%,60%);
}

.c0:-ms-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::placeholder {
  color: hsl(0,0%,60%);
}

.c0:disabled {
  cursor: not-allowed;
  background-color: hsl(0,0%,95%);
  border: 1px solid transparent;
  box-shadow: none;
}

<input
    class="c0"
    placeholder="placeholder"
  />
  notification
</DocumentFragment>
`;

exports[`Input with left addon renders correctly 1`] = `
<DocumentFragment>
  .c3 {
  height: 2.8rem;
  width: 100%;
  margin: 0.4rem 0;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 0.4rem 1.4rem;
  box-sizing: border-box;
  background-color: hsl(0,0%,95%);
  color: hsl(0,0%,5%);
  font-size: 1.4rem;
  line-height: 2rem;
}

.c3:hover {
  background-color: hsl(0,0%,90%);
  color: hsl(0,0%,5%);
}

.c3:focus,
.c3:active {
  outline: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid #5ab2f2;
  box-shadow: 0 0 2px hsl(216,100%,50%);
}

.c3::-webkit-input-placeholder {
  color: hsl(0,0%,60%);
}

.c3::-moz-placeholder {
  color: hsl(0,0%,60%);
}

.c3:-ms-input-placeholder {
  color: hsl(0,0%,60%);
}

.c3::placeholder {
  color: hsl(0,0%,60%);
}

.c3:disabled {
  cursor: not-allowed;
  background-color: hsl(0,0%,95%);
  border: 1px solid transparent;
  box-shadow: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
}

.c0 input {
  padding-left: 34px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c2 {
  cursor: inherit;
  display: inline-block;
  height: 18px;
  width: 18px;
}

<span
    class="c0"
  >
    <span
      class="c1"
    >
      <span
        class="c2"
        height="18px"
        type="oInfo"
        width="18px"
      >
        <svg
          fill="#000"
          height="18px"
          viewBox="0 0 24 24"
          width="18px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 7H13V9H11V7ZM11 11H13V17H11V11ZM12 2C6.48 2 2 6.48 2 12C2 17.52 6.48 22 12 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 12 2ZM12 20C7.59 20 4 16.41 4 12C4 7.59 7.59 4 12 4C16.41 4 20 7.59 20 12C20 16.41 16.41 20 12 20Z"
          />
        </svg>
      </span>
    </span>
    <input
      class="c3"
      placeholder="placeholder"
    />
  </span>
</DocumentFragment>
`;

exports[`Input with left and right addons renders correctly 1`] = `
<DocumentFragment>
  .c3 {
  height: 2.8rem;
  width: 100%;
  margin: 0.4rem 0;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 0.4rem 1.4rem;
  box-sizing: border-box;
  background-color: hsl(0,0%,95%);
  color: hsl(0,0%,5%);
  font-size: 1.4rem;
  line-height: 2rem;
}

.c3:hover {
  background-color: hsl(0,0%,90%);
  color: hsl(0,0%,5%);
}

.c3:focus,
.c3:active {
  outline: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid #5ab2f2;
  box-shadow: 0 0 2px hsl(216,100%,50%);
}

.c3::-webkit-input-placeholder {
  color: hsl(0,0%,60%);
}

.c3::-moz-placeholder {
  color: hsl(0,0%,60%);
}

.c3:-ms-input-placeholder {
  color: hsl(0,0%,60%);
}

.c3::placeholder {
  color: hsl(0,0%,60%);
}

.c3:disabled {
  cursor: not-allowed;
  background-color: hsl(0,0%,95%);
  border: 1px solid transparent;
  box-shadow: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
}

.c0 input {
  padding-left: 34px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c2 {
  cursor: inherit;
  display: inline-block;
  height: 18px;
  width: 18px;
}

<span
    class="c0"
  >
    <span
      class="c1"
    >
      <span
        class="c2"
        height="18px"
        type="oInfo"
        width="18px"
      >
        <svg
          fill="#000"
          height="18px"
          viewBox="0 0 24 24"
          width="18px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 7H13V9H11V7ZM11 11H13V17H11V11ZM12 2C6.48 2 2 6.48 2 12C2 17.52 6.48 22 12 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 12 2ZM12 20C7.59 20 4 16.41 4 12C4 7.59 7.59 4 12 4C16.41 4 20 7.59 20 12C20 16.41 16.41 20 12 20Z"
          />
        </svg>
      </span>
    </span>
    <input
      class="c3"
      placeholder="placeholder"
    />
  </span>
</DocumentFragment>
`;

exports[`Input with right addon renders correctly 1`] = `
<DocumentFragment>
  .c1 {
  height: 2.8rem;
  width: 100%;
  margin: 0.4rem 0;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 0.4rem 1.4rem;
  box-sizing: border-box;
  background-color: hsl(0,0%,95%);
  color: hsl(0,0%,5%);
  font-size: 1.4rem;
  line-height: 2rem;
}

.c1:hover {
  background-color: hsl(0,0%,90%);
  color: hsl(0,0%,5%);
}

.c1:focus,
.c1:active {
  outline: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid #5ab2f2;
  box-shadow: 0 0 2px hsl(216,100%,50%);
}

.c1::-webkit-input-placeholder {
  color: hsl(0,0%,60%);
}

.c1::-moz-placeholder {
  color: hsl(0,0%,60%);
}

.c1:-ms-input-placeholder {
  color: hsl(0,0%,60%);
}

.c1::placeholder {
  color: hsl(0,0%,60%);
}

.c1:disabled {
  cursor: not-allowed;
  background-color: hsl(0,0%,95%);
  border: 1px solid transparent;
  box-shadow: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
}

.c0 input {
  padding-right: 34px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c3 {
  cursor: inherit;
  display: inline-block;
  height: 18px;
  width: 18px;
}

<span
    class="c0"
  >
    <input
      class="c1"
      placeholder="placeholder"
    />
    <span
      class="c2"
    >
      <span
        class="c3"
        height="18px"
        type="oInfo"
        width="18px"
      >
        <svg
          fill="#000"
          height="18px"
          viewBox="0 0 24 24"
          width="18px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 7H13V9H11V7ZM11 11H13V17H11V11ZM12 2C6.48 2 2 6.48 2 12C2 17.52 6.48 22 12 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 12 2ZM12 20C7.59 20 4 16.41 4 12C4 7.59 7.59 4 12 4C16.41 4 20 7.59 20 12C20 16.41 16.41 20 12 20Z"
          />
        </svg>
      </span>
    </span>
  </span>
</DocumentFragment>
`;

exports[`Input with size large renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  height: 3.2rem;
  width: 100%;
  margin: 0.4rem 0;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 0.6rem 1.4rem;
  box-sizing: border-box;
  background-color: hsl(0,0%,95%);
  color: hsl(0,0%,5%);
  font-size: 1.4rem;
  line-height: 2rem;
}

.c0:hover {
  background-color: hsl(0,0%,90%);
  color: hsl(0,0%,5%);
}

.c0:focus,
.c0:active {
  outline: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid #5ab2f2;
  box-shadow: 0 0 2px hsl(216,100%,50%);
}

.c0::-webkit-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::-moz-placeholder {
  color: hsl(0,0%,60%);
}

.c0:-ms-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::placeholder {
  color: hsl(0,0%,60%);
}

.c0:disabled {
  cursor: not-allowed;
  background-color: hsl(0,0%,95%);
  border: 1px solid transparent;
  box-shadow: none;
}

<input
    class="c0"
    placeholder="placeholder"
  />
</DocumentFragment>
`;

exports[`Input with size small renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  height: 2rem;
  width: 100%;
  margin: 0.4rem 0;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 0.1rem 1.4rem;
  box-sizing: border-box;
  background-color: hsl(0,0%,95%);
  color: hsl(0,0%,5%);
  font-size: 1.2rem;
  line-height: 1.8rem;
}

.c0:hover {
  background-color: hsl(0,0%,90%);
  color: hsl(0,0%,5%);
}

.c0:focus,
.c0:active {
  outline: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid #5ab2f2;
  box-shadow: 0 0 2px hsl(216,100%,50%);
}

.c0::-webkit-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::-moz-placeholder {
  color: hsl(0,0%,60%);
}

.c0:-ms-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::placeholder {
  color: hsl(0,0%,60%);
}

.c0:disabled {
  cursor: not-allowed;
  background-color: hsl(0,0%,95%);
  border: 1px solid transparent;
  box-shadow: none;
}

<input
    class="c0"
    placeholder="placeholder"
  />
</DocumentFragment>
`;

exports[`Input with size undefined renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  height: 2.8rem;
  width: 100%;
  margin: 0.4rem 0;
  border-radius: 4px;
  border: 1px solid transparent;
  padding: 0.4rem 1.4rem;
  box-sizing: border-box;
  background-color: hsl(0,0%,95%);
  color: hsl(0,0%,5%);
  font-size: 1.4rem;
  line-height: 2rem;
}

.c0:hover {
  background-color: hsl(0,0%,90%);
  color: hsl(0,0%,5%);
}

.c0:focus,
.c0:active {
  outline: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid #5ab2f2;
  box-shadow: 0 0 2px hsl(216,100%,50%);
}

.c0::-webkit-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::-moz-placeholder {
  color: hsl(0,0%,60%);
}

.c0:-ms-input-placeholder {
  color: hsl(0,0%,60%);
}

.c0::placeholder {
  color: hsl(0,0%,60%);
}

.c0:disabled {
  cursor: not-allowed;
  background-color: hsl(0,0%,95%);
  border: 1px solid transparent;
  box-shadow: none;
}

<input
    class="c0"
    placeholder="placeholder"
  />
</DocumentFragment>
`;
